### PR TITLE
Implement strip_workspace_name.

### PR DESCRIPTION
**Pull request by c-edw**
#3082.
~~For some reason, the `strip_workspace_name` is deemed invalid in the config file. Still trying to find out why.~~

### CODE REVIEW

File: sway/commands/bar/strip_workspace_name.c [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	struct cmd_results *error = NULL;
+	if ((error = checkarg(argc,
+				"strip_workspace_name", EXPECTED_EQUAL_TO, 1))) {
+		return error;
+	}
+	if (!config->current_bar) {
+		return cmd_results_new(CMD_FAILURE,
+				"strip_workspace_name", "No bar defined.");
+	}
+	if (strcasecmp("yes", argv[0]) == 0) {
```
Comment (c-edw) [inline sway/commands/bar/strip_workspace_name.c]: Copied this part from `strip_workspace_number.c`. Might be better to use `parse_boolean` here.

Comment (SirCmpwn) [reply sway/commands/bar/strip_workspace_name.c]: +1, suggest using it in both places

**Comment by c-edw** [comment]
I'm almost certain that I have put a `strip_workspace_name` everywhere necessary, but it's still invalid when the config loads.

```
2018-11-06 20:18:29 - [sway/config.c:655] Read line 212:     strip_workspace_numbers yes
2018-11-06 20:18:29 - [sway/commands.c:374] handling config command 'bar strip_workspace_numbers yes'
2018-11-06 20:18:29 - [sway/commands.c:156] find_handler(strip_workspace_numbers)
2018-11-06 20:18:29 - [sway/commands.c:409] Subcommand: strip_workspace_numbers yes
2018-11-06 20:18:29 - [sway/commands.c:156] find_handler(strip_workspace_numbers)
2018-11-06 20:18:29 - [sway/commands/bar/strip_workspace_numbers.c:23] Stripping workspace numbers on bar: bar-0
2018-11-06 20:18:29 - [sway/config.c:655] Read line 213:     strip_workspace_name yes
2018-11-06 20:18:29 - [sway/commands.c:374] handling config command 'bar strip_workspace_name yes'
2018-11-06 20:18:29 - [sway/commands.c:156] find_handler(strip_workspace_name)
2018-11-06 20:18:29 - [sway/commands.c:409] Subcommand: strip_workspace_name yes
2018-11-06 20:18:29 - [sway/commands.c:156] find_handler(strip_workspace_name)
2018-11-06 20:18:29 - [sway/config.c:708] Error on line 213 'strip_workspace_name yes': Unknown/invalid command (/home/cedwards/.config/sway/config)
```

It fails to find the handler, even though I've defined it:
```
...
{ "strip_workspace_numbers", bar_cmd_strip_workspace_numbers },
{ "strip_workspace_name", bar_cmd_strip_workspace_name },
...
```

I've also tried an `rm -rf build` since I had to add a file to `meson.build`. No luck.

**Comment by RedSoxFan** [comment]
> ...
> { "strip_workspace_numbers", bar_cmd_strip_workspace_numbers },
> { "strip_workspace_name", bar_cmd_strip_workspace_name },
> ...

The handlers need to be in alphabetical order. Just reverse the two lines and it should work.

**Comment by c-edw** [comment]
Can't believe I missed that, thanks a lot.

File: swaybar/render.c [hunk 2]
```diff
@@ -355,17 +355,8 @@ static const char *strip_workspace_number(const char *ws_name) {
...
-		if (ws_name[i] < '0' || ws_name[i] > '9') {
-			if (':' == ws_name[i] && i < len - 1 && i > 0) {
-				return &ws_name[0];
-			}
-			return ws_name;
-		}
-	}
-	return ws_name;
+static const char *strip_workspace_name(char *ws_name) {
+	return strtok(ws_name, ":");
```
Comment (c-edw) [inline swaybar/render.c]: This is likely a poor way of doing it, but I am not familiar enough with C to know a better way...

File: sway/commands/bar/strip_workspace_name.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+				"strip_workspace_name", EXPECTED_EQUAL_TO, 1))) {
+		return error;
+	}
+	if (!config->current_bar) {
+		return cmd_results_new(CMD_FAILURE,
+				"strip_workspace_name", "No bar defined.");
+	}
+
+	config->current_bar->strip_workspace_name = 
+		parse_boolean(argv[0], config->current_bar->strip_workspace_numbers);
```
Comment (RyanDwyer) [inline sway/commands/bar/strip_workspace_name.c]: s/numbers/name

File: sway/commands/bar/strip_workspace_name.c [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+	if (config->current_bar->strip_workspace_name) {
+		wlr_log(WLR_DEBUG, "Stripping workspace name on bar: %s",
```
Comment (RyanDwyer) [inline sway/commands/bar/strip_workspace_name.c]: Should you force `strip_workspace_numbers` to `false` here?

File: sway/commands/bar/strip_workspace_numbers.c [hunk 5]
```diff
@@ -13,17 +14,17 @@ struct cmd_results *bar_cmd_strip_workspace_numbers(int argc, char **argv) {
...
+	
+	config->current_bar->strip_workspace_numbers = 
+		parse_boolean(argv[0], config->current_bar->strip_workspace_numbers);
+
+	if (config->current_bar->strip_workspace_numbers) {
 		wlr_log(WLR_DEBUG, "Stripping workspace numbers on bar: %s",
 				config->current_bar->id);
-	} else if (strcasecmp("no", argv[0]) == 0) {
-		config->current_bar->strip_workspace_numbers = false;
+	} else {
```
Comment (RyanDwyer) [inline sway/commands/bar/strip_workspace_numbers.c]: Set `strip_workspace_name` to `false` here.

File: swaybar/ipc.c [hunk 6]
```diff
@@ -153,7 +153,7 @@ static bool ipc_parse_config(
 		return false;
 	}
 	json_object *markup, *mode, *hidden_state, *position, *status_command;
-	json_object *font, *bar_height, *wrap_scroll, *workspace_buttons, *strip_workspace_numbers;
+	json_object *font, *bar_height, *wrap_scroll, *workspace_buttons, *strip_workspace_numbers, *strip_workspace_name;
```
Comment (RyanDwyer) [inline swaybar/ipc.c]: This whole function appears to ignore the 80 character line limit, but that doesn't mean you should do the same. I'm not fussed as to whether you wrap the other lines or not.

File: swaybar/render.c (hunk 2 above)
Comment (RyanDwyer) [reply swaybar/render.c]: `strtok` allocates a new string, unlike `strip_workspace_number` which returns a pointer within the same string. As it stands there's a memory leak here.

I note this is being allocated while rendering, so I think a different approach needs to be taken. We could add a new property to the `swaybar_workspace` struct for the renderable name (ie. stripped), then update it when we receive workspace events. Then render using that property.

File: swaybar/render.c [hunk 7]
```diff
@@ -371,6 +375,8 @@ static uint32_t render_workspace_button(cairo_t *cairo,
 	const char *name = ws->name;
 	if (config->strip_workspace_numbers) {
 		name = strip_workspace_number(ws->name);
+	} else if(config->strip_workspace_name) {
```
Comment (RyanDwyer) [inline swaybar/render.c]: Style: Space between `if` and `(`.

File: swaybar/render.c (hunk 2 above)
Comment (c-edw) [reply swaybar/render.c]: Should I add a `ws_number` and `ws_name` field to `swaybar_workspace` then? I assume they'd both need to be `char *`.

File: swaybar/ipc.c [hunk 8]
```diff
@@ -298,6 +304,19 @@ bool ipc_get_workspaces(struct swaybar *bar) {
 					calloc(1, sizeof(struct swaybar_workspace));
 				ws->num = json_object_get_int(num);
 				ws->name = strdup(json_object_get_string(name));
+				char *ws_dup = strdup(ws->name);
+				char *stripped_name = strtok(ws_dup, ":");
```
[the remark this answers is not in the document]
Comment (c-edw) [reply swaybar/ipc.c]: There's definitely a few holes here, like the one you mentioned, and if you renamed a workspace to something like `1:a:b`, the `stripped_name` will just be `a` and not `a:b`. I could try to fix this, but I wouldn't know how to do the validity check. I'd probably need to cast `ws->num` to a char and compare it to the first character of `ws->name` right?

[the remark this answers is not in the document]
Comment (ianyfan) [reply swaybar/ipc.c]: I don't think it needs to be that complicated. Just check if `ws->num == -1` to see if the string begins with a valid number.

[the remark this answers is not in the document]
Comment (ianyfan) [reply swaybar/ipc.c]: Another solution I thought of is to not have `stripped_name` and instead convert `ws->num` to a string. Could also have `stripped_number` just be either an index or pointer of where the name begins in `ws->name` and it'll still work as a valid string, though I'm not sure if it is more or less hassle than duplicating the string and freeing it each time.

[the remark this answers is not in the document]
Comment (c-edw) [reply swaybar/ipc.c]: I believe that `ws->num` is independent from `ws->name`. The `ws->num` workspace number is assigned automatically and cannot be changed, but the `stripped_number` is the 'frontend' number which can be changed.

[the remark this answers is not in the document]
Comment (ianyfan) [reply swaybar/ipc.c]: I don't think so: https://github.com/swaywm/sway/blob/7fa7f4f48d17e0470c800b258061d188ceb705da/sway/ipc-json.c#L191

**Comment by c-edw** [comment]
This commit fixes all the edge cases I'm aware of. I noticed some unrelated behaviour with custom workspace names - if you rename a workspace so that the `num` becomes 2 (`swaymsg rename workspace "1" to "2"`), using `$mod+2` won't switch to that workspace. Is this intended?

File: swaybar/ipc.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (c-edw) [reply swaybar/ipc.c]: Thank you for your help resolving these issues :+1:

File: swaybar/ipc.c [hunk 9]
```diff
@@ -298,6 +304,19 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
+				// ws->num will be 0 if workspace name doesn't begin with int.
+				if (stripped_name != NULL && 
+						strlen(ws->name) > 2 && ws->num != 0) {
```
Comment (ianyfan) [inline swaybar/ipc.c]: Not 0, -1.
Also, using a length of 2 misses cases when num > 9.
I think you should do `strlen(ws->name) > strlen(stripped_name) + 1`. (or set some variable to `strlen(stripped_name) + 1` and use that as the offset later too)

Comment (c-edw) [reply swaybar/ipc.c]: According to documentation, `atoi` returns 0 if invalid. Existing code already assumes that `num` is 0-9.

**Comment by c-edw** [comment]
@RyanDwyer I think this commit should do it the way you suggested.

File: swaybar/ipc.c (hunk 9 above)
Comment (ianyfan) [reply swaybar/ipc.c]: https://github.com/swaywm/sway/blob/7fa7f4f48d17e0470c800b258061d188ceb705da/sway/ipc-json.c#L191 gives -1 for num if there isn't a number.

Comment (c-edw) [reply swaybar/ipc.c]: Yeah you're right, thanks.

File: swaybar/ipc.c [hunk 10]
```diff
@@ -298,6 +304,21 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
+				char *ws_dup = strdup(ws->name);
+				char *stripped_name = strtok(ws_dup, ":");
+				size_t num_len = strlen(stripped_name) + 1;
+				// ws->num will be 0 if workspace name doesn't begin with int.
+				if (stripped_name != NULL && 
+						strlen(ws->name) > num_len && ws->num != -1) {
+					if(bar->config->strip_workspace_name) {
+						ws->label = strdup(stripped_name);
+					} else if(bar->config->strip_workspace_numbers) {
+						// Strip [n:] using offset 2.
```
Comment (ianyfan) [inline swaybar/ipc.c]: This comment needs to be changed.

File: swaybar/ipc.c [hunk 11]
```diff
@@ -298,6 +304,21 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
 				ws->name = strdup(json_object_get_string(name));
+				ws->label = strdup(ws->name);
+				char *ws_dup = strdup(ws->name);
+				char *stripped_name = strtok(ws_dup, ":");
+				size_t num_len = strlen(stripped_name) + 1;
+				// ws->num will be 0 if workspace name doesn't begin with int.
+				if (stripped_name != NULL && 
+						strlen(ws->name) > num_len && ws->num != -1) {
+					if(bar->config->strip_workspace_name) {
+						ws->label = strdup(stripped_name);
```
Comment (RedSoxFan) [inline swaybar/ipc.c]: You need to free `ws->label` before assigning a new value here to avoid a memory leak.

File: swaybar/ipc.c [hunk 12]
```diff
@@ -298,6 +304,21 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
+					} else if(bar->config->strip_workspace_numbers) {
+						// Strip number prefix [1-?:] using offset num_len.
+						ws->label = strdup(ws->name + num_len);
```
Comment (RedSoxFan) [inline swaybar/ipc.c]: `ws->label` also needs to be free'd here

File: swaybar/ipc.c (hunk 11 above)
Comment (c-edw) [reply swaybar/ipc.c]: Ah, so I need to free the pointer before reassign otherwise it will be dangling, correct?

Comment (RedSoxFan) [reply swaybar/ipc.c]: Correct. `ws->label` held the only reference to the value.

Comment (c-edw) [reply swaybar/ipc.c]: Got it, thanks.

File: swaybar/ipc.c [hunk 13]
```diff
@@ -298,6 +304,23 @@ bool ipc_get_workspaces(struct swaybar *bar) {
 					calloc(1, sizeof(struct swaybar_workspace));
 				ws->num = json_object_get_int(num);
 				ws->name = strdup(json_object_get_string(name));
+				ws->label = strdup(ws->name);
+				char *ws_dup = strdup(ws->name);
+				char *stripped_name = strtok(ws_dup, ":");
```
Comment (ianyfan) [inline swaybar/ipc.c]: I just did some testing and it seems like i3 strips names even if the colon is not present: i.e. "1a" will be stripped into 1 and a. So this logic will have to be changed if we want to match their behaviour.

Comment (c-edw) [reply swaybar/ipc.c]: I think this could be a bug. The [docs](https://i3wm.org/docs/userguide.html#_strip_workspace_numbers_name) for `strip_workspace_name` and `strip_workspace_number` emphasise that the name needs to be in the form `[n]:[NAME]` to apply.

Comment (c-edw) [reply swaybar/ipc.c]: They have confirmed that this is intended. I will see if I can get this done soon.

**Comment by c-edw** [comment]
I have managed to get `strip_workspace_numbers` to work with/without a semicolon according to i3's behaviour, but having issues with `strip_workspace_name`. 

There's some kind of memory bug I can't figure out. I'm allocating 2 bytes, one for the first character and one for the null terminator, and then calling `strncat` to copy the first character, but something is going wrong here.

File: swaybar/ipc.c [hunk 14]
```diff
@@ -305,22 +305,19 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
 				ws->label = strdup(ws->name);
-				char *ws_dup = strdup(ws->name);
-				char *stripped_name = strtok(ws_dup, ":");
-				size_t num_len = strlen(stripped_name) + 1;
-				// ws->num will be 0 if workspace name doesn't begin with int.
-				if (stripped_name != NULL && 
-						strlen(ws->name) > num_len && ws->num != -1) {
-					if(bar->config->strip_workspace_name) {
+				size_t len_offset = ws->name[1] == ':' ? 2 : 1;
+				// ws->num will be -1 if workspace name doesn't begin with int.
```
Comment (c-edw) [inline swaybar/ipc.c]: This code makes the assumption that the digit will always be a single character (0-9). Not ideal, but existing code already does this.

File: swaybar/ipc.c [hunk 15]
```diff
@@ -305,22 +305,19 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
+				if (ws->num != -1 && strlen(ws->name) > len_offset) {
+					if (bar->config->strip_workspace_name) {
 						free(ws->label);
-						ws->label = strdup(stripped_name);
-					} else if(bar->config->strip_workspace_numbers) {
+						ws->label = (char*) malloc(2 * sizeof(char));
+						strncat(ws->label, ws->name, 1);
```
Comment (ianyfan) [inline swaybar/ipc.c]: Maybe you're thinking of `strncpy`

Comment (c-edw) [reply swaybar/ipc.c]: I tried `strncpy` originally and thought that maybe the issues I've been seeing were because `strncpy` doesn't append null terminators (or does it?).

**Comment by c-edw** [comment]
Ready for review again, it has the same behaviour as i3 AFAIK from testing.

File: swaybar/ipc.c [hunk 16]
```diff
@@ -305,16 +305,18 @@ bool ipc_get_workspaces(struct swaybar *bar) {
 				ws->num = json_object_get_int(num);
 				ws->name = strdup(json_object_get_string(name));
 				ws->label = strdup(ws->name);
-				size_t len_offset = ws->name[1] == ':' ? 2 : 1;
+				size_t len_offset = numlen(ws->num);
```
Comment (ianyfan) [inline swaybar/ipc.c]: This can do weird stuff if `ws->num == -1`.

i3 also strips the colon if the name is something like `1:`. I don't think this does that, though admittedly I haven't actually tested it.

Comment (c-edw) [reply swaybar/ipc.c]: Nice catch with `numlen`, I have modified it to pass the absolute value to `log10`. Hanging colon will be stripped now too.

**Comment by c-edw** [comment]
I have done some testing and these cases now work:
- strip_workspace_name `1234:` names the workspace `1234`
- strip_workspace_numbers `:abcd` names the workspace `abcd`
- strip_workspace_name `1234:abcd` names the workspace `1234`
- strip_workspace_numbers `1234:abcd` names the workspace `abcd`
- strip_workspace_name `1234abcd` names the workspace `1234`
- strip_workspace_numbers `1234abcd` names the workspace `abcd`

**Comment by ianyfan** [comment]
strip_workspace_numbers `:abcd` should name the workspace `:abcd`

**Comment by c-edw** [comment]
@ddevault It looks like something has gone wrong with the CI.

File: swaybar/ipc.c [hunk 17]
```diff
@@ -306,18 +306,20 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
+						ws->label = malloc(len_offset + 1 * sizeof(char));
+						ws->label[len_offset] = '\0';
+						strncpy(ws->label, ws->name, len_offset);
+					} else if (bar->config->strip_workspace_numbers) {
+						len_offset += ws->label[len_offset] == ':';
+						if (strlen(ws->name) > len_offset) {
+							free(ws->label);
+							// Strip number prefix [1-?:] using len_offset.
+							ws->label = strdup(ws->name + len_offset);
+						}
```
Comment (ianyfan) [inline swaybar/ipc.c]: If `strlen(ws->name) == len_offset` this should just strip the last colon, matching the behaviour of `strip_workspace_name`.

Comment (c-edw) [reply swaybar/ipc.c]: Can you clarify further? Changing it to a `==` could cause problems because `len_offset` will always be one less than `strlen(ws->name)`.

Comment (ianyfan) [reply swaybar/ipc.c]: "1:" gets displayed as "1" if `strip_workspace_number` is enabled (though not if both are disabled).

Comment (c-edw) [reply swaybar/ipc.c]: But you said this is how it was on i3, no?

>i3 also strips the colon if the name is something like 1:. I don't think this does that, though admittedly I haven't actually tested it.

File: common/util.c [hunk 18]
```diff
@@ -23,7 +23,7 @@ int numlen(int n) {
 	if (n == 0) {
 		return 1;
 	}
-	return log10(n) + 1;
+	return log10(abs(n)) + 1;
```
Comment (ianyfan) [inline common/util.c]: Now that we're not calling this with -1 again, could you either revert this change or reject negative numbers until we have a use case for it?

File: sway/commands/bar/strip_workspace_name.c [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+	if ((error = checkarg(argc,
+				"strip_workspace_name", EXPECTED_EQUAL_TO, 1))) {
+		return error;
+	}
+	if (!config->current_bar) {
+		return cmd_results_new(CMD_FAILURE,
+				"strip_workspace_name", "No bar defined.");
+	}
+
+	config->current_bar->strip_workspace_name =
```
Comment (ianyfan) [inline sway/commands/bar/strip_workspace_name.c]: Trailing space

File: sway/commands/bar/strip_workspace_name.c [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+	if (config->current_bar->strip_workspace_name) {
+		config->current_bar->strip_workspace_numbers = false;
+
+		wlr_log(WLR_DEBUG, "Stripping workspace name on bar: %s",
+				config->current_bar->id);
+	} else {
+		wlr_log(WLR_DEBUG, "Enabling workspace name on bar: %s",
+				config->current_bar->id);
+	}
```
Comment (ianyfan) [inline sway/commands/bar/strip_workspace_name.c]: Trailing space

File: sway/commands/bar/strip_workspace_numbers.c [hunk 21]
```diff
@@ -13,17 +14,19 @@ struct cmd_results *bar_cmd_strip_workspace_numbers(int argc, char **argv) {
 		return cmd_results_new(CMD_FAILURE,
 				"strip_workspace_numbers", "No bar defined.");
 	}
-	if (strcasecmp("yes", argv[0]) == 0) {
-		config->current_bar->strip_workspace_numbers = true;
+	
+	config->current_bar->strip_workspace_numbers =
```
Comment (ianyfan) [inline sway/commands/bar/strip_workspace_numbers.c]: Trailing space

File: sway/commands/bar/strip_workspace_numbers.c [hunk 22]
```diff
@@ -13,17 +14,19 @@ struct cmd_results *bar_cmd_strip_workspace_numbers(int argc, char **argv) {
...
-	} else if (strcasecmp("no", argv[0]) == 0) {
-		config->current_bar->strip_workspace_numbers = false;
+	} else {
 		wlr_log(WLR_DEBUG, "Enabling workspace numbers on bar: %s",
 				config->current_bar->id);
-	} else {
-		return cmd_results_new(CMD_INVALID,
-				"strip_workspace_numbers", "Invalid value %s", argv[0]);
-	}
+	}
```
Comment (ianyfan) [inline sway/commands/bar/strip_workspace_numbers.c]: Trailing space

File: sway/sway-bar.5.scd [hunk 23]
```diff
@@ -50,6 +50,10 @@ Sway allows configuring swaybar in the sway configuration file.
 *workspace\_buttons* yes|no
 	Enables or disables workspace buttons on the bar. Default is _yes_.
 
+*strip\_workspace\_name* yes|no
+	If set to _yes_, then workspace name will be omitted from the workspace
```
Comment (ianyfan) [inline sway/sway-bar.5.scd]: "then workspace name" is grammatically incorrect. Sucks because the i3 devs should have named it `strip_workspace_names` but that's what we got.
I'm also not sure about saying "custom number" either. Maybe just "number".

File: swaybar/ipc.c [hunk 24]
```diff
@@ -153,8 +153,10 @@ static bool ipc_parse_config(
 		return false;
 	}
 	json_object *markup, *mode, *hidden_state, *position, *status_command;
-	json_object *font, *bar_height, *wrap_scroll, *workspace_buttons, *strip_workspace_numbers;
-	json_object *binding_mode_indicator, *verbose, *colors, *sep_symbol, *outputs;
+	json_object *font, *bar_height, *wrap_scroll, *workspace_buttons, 
+		*strip_workspace_numbers, *strip_workspace_name;
```
Comment (ianyfan) [inline swaybar/ipc.c]: Follow what was done previously and start a new line of declarations `json_object *...` instead of overhanging it.

Also remember to fix trailing spaces.

File: swaybar/ipc.c [hunk 25]
```diff
@@ -153,8 +153,10 @@ static bool ipc_parse_config(
...
+	json_object *binding_mode_indicator, *verbose, *colors, *sep_symbol, 
+		*outputs;
```
Comment (ianyfan) [inline swaybar/ipc.c]: Same as above, should probably move to the line below with `bindings`

File: swaybar/ipc.c [hunk 26]
```diff
@@ -298,6 +304,24 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
 				ws->num = json_object_get_int(num);
 				ws->name = strdup(json_object_get_string(name));
+				ws->label = strdup(ws->name);
+				// ws->num will be -1 if workspace name doesn't begin with int.
+				if(ws->num != -1) {
+					size_t len_offset = numlen(ws->num);
+					if (bar->config->strip_workspace_name) {
+						free(ws->label);
+						ws->label = malloc(len_offset + 1 * sizeof(char));
+						ws->label[len_offset] = '\0';
```
Comment (ianyfan) [inline swaybar/ipc.c]: I just realised you should be able to remove the three other lines in this block and this will still work, since it'll terminate the string in the correct place, even if there's leftover characters after it.

File: swaybar/ipc.c [hunk 27]
```diff
@@ -298,6 +304,24 @@ bool ipc_get_workspaces(struct swaybar *bar) {
...
+						strncpy(ws->label, ws->name, len_offset);
+					} else if (bar->config->strip_workspace_numbers) {
+						len_offset += ws->label[len_offset] == ':';
+						if (strlen(ws->name) > len_offset) {
+							free(ws->label);
```
Comment (ianyfan) [inline swaybar/ipc.c]: You could remove this line and use `strcpy` below, though it's less of a fun solution as the one above.

File: swaybar/ipc.c [hunk 28]
```diff
@@ -298,6 +304,24 @@ bool ipc_get_workspaces(struct swaybar *bar) {
 					calloc(1, sizeof(struct swaybar_workspace));
 				ws->num = json_object_get_int(num);
 				ws->name = strdup(json_object_get_string(name));
+				ws->label = strdup(ws->name);
+				// ws->num will be -1 if workspace name doesn't begin with int.
+				if(ws->num != -1) {
```
Comment (ianyfan) [inline swaybar/ipc.c]: Style: space after `if`.

File: common/util.c (hunk 18 above)
Comment (c-edw) [reply common/util.c]: I think it should remain like this. Negative numbers have lengths too, and before this change it would output a large negative number which could trigger a bug.

Comment (ianyfan) [reply common/util.c]: In that case, I think for negative numbers it should add an extra character to account for the '-' at the start.

**Comment by emersion** [comment]
Thanks!
